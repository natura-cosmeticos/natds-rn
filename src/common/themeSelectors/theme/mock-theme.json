{
  "avatarSizes": {
    "tiny": {
      "size": 24,
      "fontSize": ".625rem"
    },
    "small": {
      "size": 32,
      "fontSize": ".875rem"
    },
    "standard": {
      "size": 40,
      "fontSize": "1rem"
    },
    "large": {
      "size": 48,
      "fontSize": "1rem"
    },
    "huge": {
      "size": 80,
      "fontSize": "1.5rem"
    }
  },
  "buttonSizes": {
    "small": {
      "paddingTop": 8,
      "paddingRight": 8,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "minHeight": 32
    },
    "medium": {
      "paddingTop": 12,
      "paddingRight": 16,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "minHeight": 40
    },
    "large": {
      "paddingTop": 16,
      "paddingRight": 16,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "minHeight": 48
    }
  },
  "fontSize": {
    "h1": {
      "fontSize": "6rem",
      "fontWeight": 300
    },
    "h2": {
      "fontSize": "3.75rem",
      "fontWeight": 300
    },
    "h3": {
      "fontSize": "3rem",
      "fontWeight": 400
    },
    "h4": {
      "fontSize": "2.125rem",
      "fontWeight": 400
    },
    "h5": {
      "fontSize": "1.5rem",
      "fontWeight": 400
    },
    "h6": {
      "fontSize": "1.25rem",
      "fontWeight": 600
    },
    "subtitle1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "subtitle2": {
      "fontSize": "0.875rem",
      "fontWeight": 500
    },
    "body1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "body2": {
      "fontSize": "0.875rem",
      "fontWeight": 400
    },
    "button": {
      "fontSize": "0.875rem",
      "fontWeight": 600
    },
    "caption": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    },
    "overline": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    }
  },
  "iconSizes": {
    "micro": 16,
    "tiny": 24,
    "small": 32,
    "standard": 48,
    "medium": 64,
    "large": 72,
    "largex": 96,
    "largexx": 128,
    "huge": 144,
    "hugex": 192,
    "hugexx": 256
  },
  "opacity": {
    "12": 0.12,
    "16": 0.16,
    "24": 0.24,
    "32": 0.32,
    "48": 0.48,
    "56": 0.56,
    "64": 0.64,
    "80": 0.8,
    "00": 0,
    "04": 0.04,
    "08": 0.08
  },
  "overlay": {
    "solidBlack": 0.56
  },
  "radius": {
    "none": 0,
    "circle": "50%",
    "small": 2,
    "medium": 4,
    "large": 8
  },
  "sizes": {
    "none": 0,
    "micro": 4,
    "tiny": 8,
    "small": 16,
    "standard": 24,
    "semi": 32,
    "semix": 40,
    "medium": 48,
    "mediumx": 56,
    "large": 64,
    "largex": 72,
    "largexx": 80,
    "largexxx": 88,
    "huge": 96,
    "hugex": 128,
    "hugexx": 144,
    "hugexxx": 192,
    "veryhuge": 256
  },
  "spacing": {
    "spacing": 8,
    "spacingNone": 0,
    "spacingMicro": 4,
    "spacingTiny": 8,
    "spacingSmall": 16,
    "spacingStandard": 24,
    "spacingSemi": 32,
    "spacingLarge": 48,
    "spacingHuge": 64
  },
  "colorTokens": {
    "colorPrimary": "#F4AB34",
    "colorOnPrimary": "#333333",
    "colorPrimaryLight": "#FEFDE8",
    "colorOnPrimaryLight": "#333333",
    "colorPrimaryDark": "#EF8426",
    "colorOnPrimaryDark": "#333333",
    "colorSecondary": "#FF6B0B",
    "colorOnSecondary": "#000000",
    "colorSecondaryLight": "#FFF8E1",
    "colorOnSecondaryLight": "#000000",
    "colorSecondaryDark": "#FF5808",
    "colorOnSecondaryDark": "#000000",
    "colorBackground": "#FAFAFA",
    "colorOnBackground": "#333333",
    "colorSurface": "#FFFFFF",
    "colorOnSurface": "#333333",
    "colorHighlight": "#000000",
    "colorHighEmphasis": "#333333",
    "colorMediumEmphasis": "#777777",
    "colorLowEmphasis": "#BBBBBB",
    "colorLink": "#227BBD",
    "colorOnLink": "#FFFFFF",
    "colorSuccess": "#569A32",
    "colorOnSuccess": "#FFFFFF",
    "colorWarning": "#FCC433",
    "colorOnWarning": "#333333",
    "colorAlert": "#E74627",
    "colorOnAlert": "#FFFFFF"
  },
  "elevation": {
    "1": {
      "elevation": 1,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 1, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 1
    },
    "2": {
      "elevation": 2,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 2, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 2
    },
    "3": {
      "elevation": 3,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 3, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 4
    },
    "4": {
      "elevation": 4,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 4, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 5
    },
    "5": {
      "elevation": 5,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 6, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "6": {
      "elevation": 6,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 8, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "7": {
      "elevation": 7,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 9, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 12
    },
    "8": {
      "elevation": 8,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 12, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 17
    },
    "9": {
      "elevation": 9,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 16, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 24
    },
    "10": {
      "elevation": 10,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 24, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 38
    }
  }
}
