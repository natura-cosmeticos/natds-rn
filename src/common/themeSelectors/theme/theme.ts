import { tokens as styleTokens } from '@naturacosmeticos/natds-styles';
import { IColorThemeTokens } from '@naturacosmeticos/natds-styles/dist/tokens/colors/themeBound/IColorThemeTokens';
import { IElevation } from '@naturacosmeticos/natds-styles/dist/tokens/elevation/IElevation';

type Tokens = typeof styleTokens;
type TokensWithoutColor = Omit<Tokens, 'colors'>
type TokensWithoutColorTokens = Omit<TokensWithoutColor, 'colorTokens'>
type TokensWithoutElevation = Omit<TokensWithoutColorTokens, 'elevation'>

interface ElevationProps {
  elevation: number
  shadowColor: string
  shadowOffset: { height: number, width: number }
  shadowOpacity: number
  shadowRadius: number
}

type Elevation = Record<keyof IElevation, ElevationProps>


export type Theme = TokensWithoutElevation
  & {
    colorTokens: IColorThemeTokens,
    elevation: Elevation | {}
  }

type ThemeMode = 'dark' | 'light';

const parseElevation = (elevation: string) => {
  const regEx = /(\d+(\.\d+)?)/g;
  const matches = elevation.match(regEx);

  if (matches) {
    const [width, height, blurRadius, , r, g, b, alpha] = matches;

    const toUnit = (prop: string) => parseInt(prop, 0);

    return {
      elevation: 2,
      shadowColor: `rgb(${r}, ${g}, ${b})`,
      shadowOffset: { height: toUnit(height), width: toUnit(width) },
      shadowOpacity: parseFloat(alpha),
      shadowRadius: toUnit(blurRadius),
    };
  }

  return {};
};

const buidElevation = (elevation: IElevation) => {
  const nativeElevation = {};

  Object
    .keys(elevation)
    .forEach((key) => { nativeElevation[key] = parseElevation(elevation[key]); });

  return nativeElevation;
};

export const buildTheme = (tokens: Tokens, brand: string, mode: ThemeMode) => {
  const colorThemeName = `${brand}${mode === 'light' ? 'Light' : 'Dark'}ColorTokens`;
  const localTokens = { ...tokens };

  delete localTokens.colors;

  const theme: Theme = {
    ...localTokens,
    colorTokens: localTokens.colorTokens[colorThemeName],
    elevation: buidElevation(localTokens.elevation),
  };

  return theme;
};

interface ThemeSelector {
  (theme: Theme): string | undefined
}

export const getTheme = (selector: ThemeSelector) => (
  { theme }: { theme: Theme },
) => selector(theme);
