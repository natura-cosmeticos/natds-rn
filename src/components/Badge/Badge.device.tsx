import { storiesOf } from '@storybook/react-native'
import { All } from './Badge.stories'

storiesOf('Badge', module)
  .add('All', All)
