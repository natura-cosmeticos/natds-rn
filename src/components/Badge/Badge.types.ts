/* eslint-disable max-len */
import { Color } from '@naturacosmeticos/natds-themes/react-native'
import { AccessibilityRole, AccessibilityValue, Animated } from 'react-native'

export type AnimationProps = {
  currentValue: Animated.Value;
  duration: number;
  toValue: number;
}

export type BadgeColors = keyof Pick<Color, 'primary' | 'secondary' | 'success' | 'alert'>

export interface BadgeBaseProps {
  /**
   * Indicates that the view is an accessibility element.
   */
  accessible?: boolean;
  /**
   * When a view is marked as accessible, it is a good practice to set an accessibilityLabel on the view.
   */
  accessibilityLabel?: string;
  /**
   * Communicates the purpose of a component to the user of an assistive technology.
   */
  accessibilityRole?: AccessibilityRole;
  /**
   * Represents the current value of a component. It can be a textual description of a component's value, or for
   * range-based components, it contains range information (minimum, current, and maximum).
   */
  accessibilityValue?: AccessibilityValue;
  /**
   * Specify the color of the Badge, from a list of available colors.
   * If no color is specified, the 'alert' color will be applied as default.
   * @default 'alert'
   */
  color?: BadgeColors;
  /**
   * Optional ID for testing
   */
  testID?: string;
}

export type BadgePulseProps = BadgeBaseProps & {
  /**
   * Specify the variant of Badge you want to create.
   */
  variant: 'pulse';
}

export type BadgeDotProps = BadgeBaseProps & {
  /**
   * Specify the variant of Badge you want to create.
   */
  variant: 'dot';
}

export type BadgeStandardProps = BadgeBaseProps & {
  /**
   * Specify the limit of notification and when is equal or less than value,
   * will display a `+` sign next to the value. Only available for `standard` variant
   */
  limit?: number;
  /**
   * Specify the number that will be display inside the Badge.
   * Only available for `standard` variant
   */
  value: number;
  /**
   * Specify the variant of Badge you want to create.
   */
  variant: 'standard';
}

export type BadgeProps = BadgePulseProps | BadgeDotProps | BadgeStandardProps
