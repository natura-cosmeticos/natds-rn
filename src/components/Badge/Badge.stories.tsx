import React from 'react'
import { Badge } from './Badge'
import { StoryContainer } from '../../common/HelperComponents/StoryContainer'

const description = () => `
- - -

  ### NOTE:
  This component is available in the following variants:

  - ✅ Dot
  - ✅ Pulse
  - ✅ Standard

With the following attribute status:

  - **Color:**
    - ✅ \`Alert\`
    - ✅ \`Primary\`
    - ✅ \`Secondary\`
    - ✅ \`Success\`

  - ✅ **Limit**
  - ✅ **Value**
- - -
`

export default {
  component: Badge,
  parameters: {
    componentSubtitle: 'Badge is a visual indicator for numeric values.',
    docs: {
      extractComponentDescription: description
    }
  },
  title: 'Components/Badge'
}

export const All = () => (
  <StoryContainer title="">
    <Badge variant="dot" />
    <Badge variant="pulse" color="success" />
    <Badge variant="standard" color="primary" value={10} limit={9} />
    <Badge variant="standard" color="secondary" value={100} limit={99} />
    <Badge variant="standard" value={1000} />
  </StoryContainer>
)
