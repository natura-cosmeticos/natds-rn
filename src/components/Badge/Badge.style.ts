import styled from 'styled-components/native';

import { BadgeColors, BadgeTypes } from './Badge';

interface ContainerProps {
  type: BadgeTypes;
  color: BadgeColors;
}

interface ContentProps {
  color: BadgeColors;
}

const getBackgroundColor = (color: BadgeColors) => {
  switch (color) {
    case 'alert':
      return '#e74627';
    case 'primary':
      return '#f4ab34';
    case 'secondary':
      return '#ff6b0b';
    case 'success':
      return '#569a32';
    default:
      return '#e74627';
  }
};

const getFontColor = (color: BadgeColors) => {
  if (color === 'primary' || color === 'secondary') {
    return '#333';
  }

  return '#fff';
};

export const BadgeContainer = styled.View<ContainerProps>`
  border-radius: 50px;
  min-width: 10px;
  height: 10px;

  padding-vertical: ${({ type }) => (type === 'standard' ? 8 : 0)}px;
  padding-horizontal: ${({ type }) => (type === 'standard' ? 5 : 0)}px;

  background-color: ${({ color }) => getBackgroundColor(color)};

  justify-content: center;
  align-items: center;
`;

export const BadgeContent = styled.Text<ContentProps>`
  font-size: 12px;
  color: ${({ color }) => getFontColor(color)};
`;
