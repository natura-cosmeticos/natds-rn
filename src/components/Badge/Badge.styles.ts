import { Theme } from '@naturacosmeticos/natds-themes/react-native';
import styled from 'styled-components/native';
import { CSSObject } from 'styled-components';

import { BadgeColors } from './Badge.types';

type BadgeStyleProps = {
  color: BadgeColors
  variant: BadgeVariant
}

export type BadgeVariant = 'standard' | 'pulse' | 'dot'
export const convertPulseToDot = (variant: BadgeVariant) => (variant === 'pulse' ? 'dot' : variant);

export const getHorizontalPadding = (variant: BadgeVariant, theme: Theme) => (
  variant === 'standard' && {
    paddingHorizontal: theme.spacing.micro,
  }
);

export const getWidthByVariant = (variant: BadgeVariant, theme: Theme) => (
  variant !== 'standard' && {
    width: theme.badge[convertPulseToDot(variant)].height,
  }
);

export const Container = styled.View(() => ({
  alignContent: 'center',
  alignItems: 'center',
}));

export const BadgeBase = styled.View<BadgeStyleProps>(({ theme, color, variant = 'standard' }): CSSObject => ({
  ...getWidthByVariant(variant, theme),
  backgroundColor: theme.badge.color[color].background,
  borderRadius: theme.badge[convertPulseToDot(variant)].borderRadius,
  height: theme.badge[convertPulseToDot(variant)].height,
}));

export const Circle = styled(BadgeBase)(({ theme, variant = 'standard' }): CSSObject => ({
  ...getHorizontalPadding(variant, theme),
  alignContent: 'center',
  justifyContent: 'center',
}));

export const Label = styled.Text<BadgeStyleProps>(({ theme, color, variant = 'standard' }): CSSObject => ({
  color: theme.badge.color[color].label,
  fontFamily: theme.badge.label.primary.fontFamily,
  fontSize: theme.badge.label.fontSize,
  fontWeight: theme.badge.label.primary.fontWeight,
  letterSpacing: theme.badge.label.letterSpacing,
  lineHeight: theme.badge[convertPulseToDot(variant)].height,
  textAlignVertical: 'center',
}));
