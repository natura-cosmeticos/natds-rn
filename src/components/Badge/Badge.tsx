/* eslint-disable max-len */
import React, { useEffect, useRef } from 'react';
import { Animated, Platform } from 'react-native';

import {
  Container, Circle, BadgeBase as Pulse, Label,
} from './Badge.styles';
import { BadgeProps, BadgeStandardProps, AnimationProps } from './Badge.types';

export const buildValue = ({ value, limit }: BadgeStandardProps) => (limit && value >= limit ? `${limit}+` : value);
export const isBadgeStandard = (props: BadgeProps): props is BadgeStandardProps => props.variant === 'standard';

export const Badge = (props: BadgeProps) => {
  const scale = useRef(new Animated.Value(1)).current;
  const opacity = useRef(new Animated.Value(1)).current;
  const animationTime = 1600;

  const {
    color = 'alert', variant, testID = 'ds-badge', accessible, accessibilityLabel, accessibilityRole, accessibilityValue,
  } = props;

  const setAnimation = ({ currentValue, duration, toValue }: AnimationProps) => {
    const animation = Animated.timing(currentValue, {
      duration,
      isInteraction: false,
      toValue,
      useNativeDriver: Platform.OS !== 'web',
    });

    Animated.loop(animation).start();
  };

  useEffect(() => {
    setAnimation({ currentValue: scale, duration: animationTime, toValue: 2.25 });
    setAnimation({ currentValue: opacity, duration: animationTime, toValue: 0 });
  }, []);

  return (
    <Container
      testID={testID}
      accessible={accessible}
      accessibilityLabel={accessibilityLabel}
      accessibilityRole={accessibilityRole}
      accessibilityValue={accessibilityValue}
      >
      <Circle color={color} variant={variant} >
        {isBadgeStandard(props) && <Label color={color} variant={variant}>{buildValue(props)}</Label>}
      </Circle>
      {variant === 'pulse' && <Pulse testID='ds-badge-pulse' as={Animated.View} color={color} variant={variant} style={{
        opacity, position: 'absolute', transform: [{ scale }], zIndex: -1,
      }} />}
    </Container>
  );
};
