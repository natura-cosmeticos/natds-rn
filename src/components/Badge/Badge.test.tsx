import React from 'react';

import { renderWithTheme } from '../../../test/testHelpers';
import {
  Badge, buildValue, isBadgeStandard,
} from '.';
import { BadgeProps } from './Badge.types';
import {
  convertPulseToDot, getWidthByVariant, getHorizontalPadding,
} from './Badge.styles';
import theme from '../../common/themeSelectors/theme/mock-theme.json';

jest.useFakeTimers();

describe('Badge component', () => {
  it('Should render badge with standard variant', () => {
    const { toJSON } = renderWithTheme(<Badge variant="standard" value={1} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('Should render badge with pulse variant', () => {
    const { getByTestId } = renderWithTheme(<Badge variant="pulse" />);

    expect(getByTestId('ds-badge-pulse')).toBeTruthy();
  });
});

describe('getWidthByVariant', () => {
  it('should only return the value when variant is different than stardard', () => {
    expect(getWidthByVariant('standard', theme)).toBe(false);
    expect(getWidthByVariant('dot', theme)).toStrictEqual({ width: 8 });
  });
});

describe('getHorizontalPadding', () => {
  it('should only return the value when variant is equal standard', () => {
    expect(getHorizontalPadding('standard', theme)).toStrictEqual({ paddingHorizontal: 4 });
    expect(getHorizontalPadding('dot', theme)).toBe(false);
  });
});

describe('convertBadgeType', () => {
  it('should return dot when pulse', () => {
    expect(convertPulseToDot('pulse')).toBe('dot');
  });

  it('should return the own variant when different than pulse', () => {
    expect(convertPulseToDot('dot')).toBe('dot');
    expect(convertPulseToDot('standard')).toBe('standard');
  });
});

describe('buildValue', () => {
  it('should only return the value when it is less than the limit', () => {
    expect(buildValue({ limit: 100, value: 99, variant: 'standard' })).toBe(99);
  });
  it('should return the limit with a plus sign when the value is greater or equal than it', () => {
    expect(buildValue({ limit: 99, value: 99, variant: 'standard' })).toBe('99+');
  });
});

describe('isBadgeStandard', () => {
  it('should return true when the variant is standard', () => {
    const props = { variant: 'standard' } as BadgeProps;

    expect(isBadgeStandard(props)).toBe(true);
  });

  it('should return false when the variant is different than standard', () => {
    const props = { variant: 'dot' } as BadgeProps;

    expect(isBadgeStandard(props)).toBe(false);
  });
});
