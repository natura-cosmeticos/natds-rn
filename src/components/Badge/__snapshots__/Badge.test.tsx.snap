// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge component Should render badge with standard variant 1`] = `
<View
  style={
    Array [
      Object {
        "alignContent": "center",
        "alignItems": "center",
      },
    ]
  }
  testID="ds-badge"
>
  <View
    color="alert"
    style={
      Array [
        Object {
          "alignContent": "center",
          "backgroundColor": "#e74627",
          "borderRadius": 8,
          "height": 16,
          "justifyContent": "center",
          "paddingHorizontal": 4,
        },
      ]
    }
    variant="standard"
  >
    <Text
      color="alert"
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontFamily": "Roboto",
            "fontSize": 12,
            "fontWeight": "400",
            "letterSpacing": 0.38,
            "lineHeight": 16,
            "textAlignVertical": "center",
          },
        ]
      }
      variant="standard"
    >
      1
    </Text>
  </View>
</View>
`;
